Add support for VisionOS

ChuanlongChen opened the pull request:
# Add support for VisionOS
1. Make "sign in with Apple " works in VisionOS.
2. Verify:
already test it both in VisionOS simulator and VisionOS Device, below is the record video in the simulator
https://github.com/ChuanlongChen/apple-signin-unity/assets/87799934/d269bb52-c0e3-4b6d-b35f-28d2982ac4e3

3. Environment:
(1) UnityEditor Version:  2022.3.18 f1
(2) Xcode Version 15.2 beta

+@lupidan  to help review

lupidan commented on README.md:
```diff
@@ -130,7 +132,7 @@ openupm add com.lupidan.apple-signin-unity
 
 ![Import detail](./Img/ImportPlugin.png)
 
-## Plugin setup (iOS/tvOS)
```
Let's have a small visionOS section instead

ChuanlongChen replied:
done

lupidan commented on README.md:
```diff
@@ -146,32 +148,51 @@ The provided extension method is `AddSignInWithAppleWithCompatibility`. It accep
+    [PostProcessBuild(CallOrder)]
+    public static void OnPostProcessBuild(BuildTarget target, string path) {
+      if (target == BuildTarget.iOS || target == BuildTarget.tvOS || target == BuildTarget.VisionOS) {
+#if UNITY_XCODE_EXTENSIONS_AVAILABLE
+        string projectPath = PBXProject.GetPBXProjectPath(path);
+        if (target == BuildTarget.VisionOS) {
+          // Unity's bug:
+          // After switch to VisionOS platform the projectPath is still "xx/Unity-iPhone.xcodeproj/project.pbxproj",
+          // while the expected path is "xx/Unity-VisionOS.xcodeproj/project.pbxproj",
+          // temporary fix:
+          projectPath = projectPath.Replace("Unity-iPhone.xcodeproj", "Unity-VisionOS.xcodeproj");
+        }
```
The current adoption for visionOS is not that big.
I would rather have this code in a separate section, just before the macOS Plugin setup section and leave this documentation for iOS/tvOS as it was.

```
## Plugin setup (visionOS)

On visionOS, the setup is pretty much the same as the iOS/tvOS setup.
However, due to how Unity's visionOS project is exported, you need to adapt the Postprocess code to find the proper xcode project filename:

if (target == BuildTarget.VisionOS)
{
    projectPath = projectPath.Replace("Unity-iPhone.xcodeproj", "Unity-VisionOS.xcodeproj");
}
```

ChuanlongChen replied:
done

lupidan commented on AppleAuthSampleProject/Assets/AppleAuthSample/Editor/SignInWithApplePostprocessor.cs:
```diff
@@ -1,44 +1,46 @@
-                        manager.AddSignInWithAppleWithCompatibility(project.GetUnityFrameworkTargetGuid());
-                        manager.WriteToFile();
-                    #else
+    [PostProcessBuild(CallOrder)]
+    public static void OnPostProcessBuild(BuildTarget target, string path) {
+      if (target == BuildTarget.iOS || target == BuildTarget.tvOS || target == BuildTarget.VisionOS) {
+#if UNITY_XCODE_EXTENSIONS_AVAILABLE
+        string projectPath = PBXProject.GetPBXProjectPath(path);
+        if (target == BuildTarget.VisionOS) {
+            
+          // This is a temporary fix for the Unity Editor's bug:
+          // After switch to VisionOS platform the projectPath is still "xx/Unity-iPhone.xcodeproj/project.pbxproj",
+          // while the expected path is "xx/Unity-VisionOS.xcodeproj/project.pbxproj",
+          projectPath = projectPath.Replace("Unity-iPhone.xcodeproj", "Unity-VisionOS.xcodeproj");
+        }
+#if UNITY_2019_3_OR_NEWER
+        var project = new PBXProject();
+        project.ReadFromString(System.IO.File.ReadAllText(projectPath));
+        var manager = new ProjectCapabilityManager(projectPath, "Entitlements.entitlements", null, project.GetUnityMainTargetGuid());
+        manager.AddSignInWithAppleWithCompatibility(project.GetUnityFrameworkTargetGuid());
+        manager.WriteToFile();
+#else
```
Most of these changes are not justified. Just add the required check for VisionOS. in the current code.
Also, we need to protect the VisionOS definition, since it's does not exist in other Unity versions.
From what I can see, VisionOS module was introduced in Unity 2022.3.5f1, so let's put 2022.3 as the basis

```
#if UNITY_2022_3_OR_LATER
    if (target == BuildTarget.VisionOS)
    {
        projectPath = projectPath.Replace("Unity-iPhone.xcodeproj", "Unity-VisionOS.xcodeproj");
    }
#endif
```

ChuanlongChen replied:
done,

ChuanlongChen replied:
To achieve this, i separate the VisionOS part with the rest

ChuanlongChen replied:
verified by build and run in simulator

lupidan commented on AppleAuthSampleProject/Assets/AppleAuthSample/Editor/SignInWithApplePostprocessor.cs:
```diff
@@ -1,44 +1,46 @@
 using AppleAuth.Editor;
 using UnityEditor;
 using UnityEditor.Callbacks;
+using UnityEngine;
```
Not needed

ChuanlongChen replied:
done

lupidan commented on AppleAuthSampleProject/Assets/AppleAuthSample/Editor/SignInWithApplePostprocessor.cs:
```diff
@@ -1,44 +1,46 @@
-    public static class SignInWithApplePostprocessor
-    {
-        private const int CallOrder = 1;
+namespace AppleAuthSample.Editor {
+  public static class SignInWithApplePostprocessor {
+    private const int CallOrder = 1;
```
Dont change the original formatting

ChuanlongChen replied:
done

lupidan commented on AppleAuthSampleProject/Assets/AppleAuthSample/Editor/SignInWithApplePostprocessor.cs:
```diff
@@ -1,44 +1,46 @@
-                AppleAuthMacosPostprocessorHelper.FixManagerBundleIdentifier(target, path);
-            }
-        }
+#endif
+#endif
+      } else if (target == BuildTarget.StandaloneOSX) {
+        AppleAuthMacosPostprocessorHelper.FixManagerBundleIdentifier(target, path);
+      }
```
Don't change the original formatting

ChuanlongChen replied:
done, recoverd

lupidan commented on AppleAuth/Native/iOS/AppleAuthSerializer.m:
```diff
@@ -74,8 +74,9 @@ + (NSDictionary *) loginResponseDictionaryForAppleIdCredentialDictionary:(NSDict
 
-// IOS/TVOS 9.0 | MACOS 10.11
-#if __IPHONE_OS_VERSION_MAX_ALLOWED >= 90000 || __TV_OS_VERSION_MAX_ALLOWED >= 90000 || __MAC_OS_X_VERSION_MAX_ALLOWED >= 101100
+
```
Unrequired newline

ChuanlongChen replied:
done

lupidan commented on AppleAuth/Native/iOS/AppleAuthManager.m:
```diff
@@ -326,15 +326,17 @@ - (void) authorizationController:(ASAuthorizationController *)controller didComp
     #else
         return nil;
     #endif
+    
```
unrequired newline

ChuanlongChen replied:
done

lupidan reviewed (changes requested):
Thanks a lot for the investigation, and for taking the time to make an official Pull Request. Is stuff like this that really helps out.

I finally found some time to dive into it, and I was able to do a quick test of the solution on a friends M1 macbook. Seems to work great 👍 

Just left a few comments. Let's prepare it on the canary branch to prepare a release.

I would also like to know what would happen to those meta file changes in previous versions of Unity. Although reached this point in time, I would rather take the opportunity to put the base Unity Version at least to 2020.3 (instead of the old 2018)

ChuanlongChen commented:
> Thanks a lot for the investigation, and for taking the time to make an official Pull Request. Is stuff like this that really helps out.
> 
> I finally found some time to dive into it, and I was able to do a quick test of the solution on a friends M1 macbook. Seems to work great 👍
> 
> Just left a few comments. Let's prepare it on the canary branch to prepare a release.
> 
> I would also like to know what would happen to those meta file changes in previous versions of Unity. Although reached this point in time, I would rather take the opportunity to put the base Unity Version at least to 2020.3 (instead of the old 2018)


Hi, thank you for taking the time to review this PR, and I've just addressed the comments you left above.
What I've done to this meta file is that I've selected the "Vision OS" platform for those .h/.m files in the Unity 2022.3.18f1 Editor.

lupidan reviewed (approved):
Thanks a lot for the changes!! And thanks a lot for providing a solution for VisionOS.

I'll merge it to the canary branch to prepare a new version and eventually make a release.

![https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExeG8zcnZ2aHhpYW1xZ3YyNjd0MjdqdjhlYjZsazUybHRhYmdkbXc5MyZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/U0DnCjWB3kTfRcucxK/giphy-downsized.gif](https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExeG8zcnZ2aHhpYW1xZ3YyNjd0MjdqdjhlYjZsazUybHRhYmdkbXc5MyZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/U0DnCjWB3kTfRcucxK/giphy-downsized.gif)

ChuanlongChen commented:
Cool! Thanks for your effort!
